### PR TITLE
Add APIs for digital twin service support

### PR DESCRIPTION
**Pull request by abhipsaMisra**
APIs added:

```csharp
namespace Microsoft.Azure.Devices {
     public class DigitalTwinClient : IDisposable {
         public static DigitalTwinClient CreateFromConnectionString(string connectionString);
         public void Dispose();
         protected virtual void Dispose(bool disposing);
         public Task<HttpOperationResponse<T, DigitalTwinGetHeaders>> GetAsync<T>(string digitalTwinId, CancellationToken cancellationToken = default(CancellationToken));
         public Task<HttpOperationResponse<DigitalTwinCommandResponse, DigitalTwinInvokeCommandHeaders>> InvokeCommandAsync(string digitalTwinId, string commandName, string payload, DigitalTwinInvokeCommandRequestOptions requestOptions = null, CancellationToken cancellationToken = default(CancellationToken));
         public Task<HttpOperationResponse<DigitalTwinCommandResponse, DigitalTwinInvokeCommandHeaders>> InvokeComponentCommandAsync(string digitalTwinId, string componentName, string commandName, string payload, DigitalTwinInvokeCommandRequestOptions requestOptions = null, CancellationToken cancellationToken = default(CancellationToken));
         public Task<HttpOperationHeaderResponse<DigitalTwinUpdateHeaders>> UpdateAsync(string digitalTwinId, string digitalTwinUpdateOperations, DigitalTwinUpdateRequestOptions requestOptions = null, CancellationToken cancellationToken = default(CancellationToken));
     }
     public class DigitalTwinCommandResponse {
         public DigitalTwinCommandResponse();
         public string Payload { get; set; }
         public int? Status { get; set; }
     }
     public class DigitalTwinGetHeaders {
         public DigitalTwinGetHeaders();
         public DigitalTwinGetHeaders(string eTag = null);
         public string ETag { get; set; }
     }
     public class DigitalTwinInvokeCommandHeaders {
         public DigitalTwinInvokeCommandHeaders();
         public string RequestId { get; set; }
     }
     public class DigitalTwinInvokeCommandRequestOptions {
         public DigitalTwinInvokeCommandRequestOptions();
         public int? ConnectTimeoutInSeconds { get; set; }
         public int? ResponseTimeoutInSeconds { get; set; }
     }
     public class DigitalTwinUpdateHeaders {
         public DigitalTwinUpdateHeaders();
         public DigitalTwinUpdateHeaders(string eTag = null, string location = null);
         public string ETag { get; set; }
         public string Location { get; set; }
     }
     public class DigitalTwinUpdateRequestOptions {
         public DigitalTwinUpdateRequestOptions();
         public string IfMatch { get; set; }
     }
}

namespace Microsoft.Azure.Devices.Serialization {
    public class BasicDigitalTwin {
        public BasicDigitalTwin();
        public IDictionary<string, object> CustomProperties { get; set; }
        public string Id { get; set; }
        public DigitalTwinMetadata Metadata { get; set; }
    }
    public class DigitalTwinMetadata {
        public DigitalTwinMetadata();
        public string ModelId { get; set; }
        public IDictionary<string, object> WriteableProperties { get; }
    }
    public class UpdateOperationsUtility {
        public UpdateOperationsUtility();
        public void AppendAddOp(string path, object value);
        public void AppendRemoveOp(string path);
        public void AppendReplaceOp(string path, object value);
        public string Serialize();
    }
    public class WritableProperty {
        public WritableProperty();
        public int AckCode { get; set; }
        public string AckDescription { get; set; }
        public int AckVersion { get; set; }
        public object DesiredValue { get; set; }
        public int DesiredVersion { get; set; }
        public DateTimeOffset LastUpdateTime { get; set; }
    }
}
```

### CODE REVIEW

File: iothub/service/src/DigitalTwin/DigitalTwinInvokeCommandRequestOptions.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    /// <summary>
+    /// General request options that are applicable, but optional, for invoke command APIs.
+    /// </summary>
+    public class DigitalTwinInvokeCommandRequestOptions
+    {
+        /// <summary>
+        /// The time (in seconds) that the service waits for the device to come online.
+        /// The default is 0 seconds (which means the device must already be online) and the maximum is 300 seconds.
+        /// </summary>
+        public int ConnectTimeoutInSeconds { get; set; }
```
Comment (vinagesh) [inline iothub/service/src/DigitalTwin/DigitalTwinInvokeCommandRequestOptions.cs:15]: I think these should be nullable?

Comment (abhipsaMisra) [reply iothub/service/src/DigitalTwin/DigitalTwinInvokeCommandRequestOptions.cs:15]: The PL defines it as nullable, so I'll update the request options to be the same

File: iothub/service/src/DigitalTwin/DigitalTwinClient.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                Body = new DigitalTwinCommandResponse { Status = response.Headers.XMsCommandStatuscode, Payload = response.Body },
+                Headers = new DigitalTwinInvokeCommandHeaders { RequestId = response.Headers.XMsRequestId },
+                Request = response.Request,
+                Response = response.Response,
+            };
+        }
+
+        /// <summary>
+        /// Invoke a command on a digital twin asynchronously.
```
Comment (barustum) [inline iothub/service/src/DigitalTwin/DigitalTwinClient.cs:96]: probably reword to 'Invoke a command on a digital twin component asynchronously.'

Comment (abhipsaMisra) [reply iothub/service/src/DigitalTwin/DigitalTwinClient.cs:96]: will do!

File: e2e/test/iothub/service/DigitalTwinClientE2ETests.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                {
+                    Logger.Trace($"{nameof(DigitalTwinWithComponentOperationsAsync)}: DesiredProperty update received: {patch}, {context}");
+                    return Task.FromResult(true);
+                }, deviceClient);
+
+                // Update the root-level property "targetTemperature".
+                string propertyName = "targetTemperature";
+                double propertyValue = new Random().Next(0, 100);
+                var ops = new UpdateOperationsUtility();
+                ops.AppendAddOp($"/{propertyName}", propertyValue);
```
Comment (drwill-ms) [inline e2e/test/iothub/service/DigitalTwinClientE2ETests.cs:69]: Does it work without the root slash?

File: iothub/service/src/DigitalTwin/Authentication/SharedAccessKeyCredentials.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace Microsoft.Azure.Devices.Authentication
+{
+    internal class SharedAccessKeyCredentials : IotServiceClientCredentials
+    {
+        // Time buffer before expiry when the token should be renewed, expressed as a percentage of the time to live.
+        // The token will be renewed when it has 15% or less of the sas token's lifespan left.
+        private const int RenewalTimeBufferPercentage = 15;
+
+        private readonly object _lock = new object();
```
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/Authentication/SharedAccessKeyCredentials.cs:15]: `_sasLock`?

File: iothub/service/src/DigitalTwin/Customized/DigitalTwinUpdateRequestOptions.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    /// <summary>
+    /// General request options that are applicable, but optional, for update APIs.
+    /// </summary>
+    public class DigitalTwinUpdateRequestOptions
+    {
+        /// <summary>
+        /// A string representing a weak ETag for the entity that this request performs an operation against, as per RFC7232.
+        /// </summary>
+        /// <remarks>
+        /// <para>
```
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/Customized/DigitalTwinUpdateRequestOptions.cs:15]: para tag not needed

File: e2e/test/iothub/service/DigitalTwinClientE2ETests.cs (hunk 3 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/service/DigitalTwinClientE2ETests.cs:69]: not sure, let me try it out

File: iothub/service/src/DigitalTwin/Authentication/SharedAccessKeyCredentials.cs (hunk 4 above)
Comment (abhipsaMisra) [reply iothub/service/src/DigitalTwin/Authentication/SharedAccessKeyCredentials.cs:15]: sure!

File: iothub/service/src/DigitalTwin/Customized/DigitalTwinUpdateRequestOptions.cs (hunk 5 above)
Comment (abhipsaMisra) [reply iothub/service/src/DigitalTwin/Customized/DigitalTwinUpdateRequestOptions.cs:15]: will remove it then

File: e2e/test/iothub/service/DigitalTwinClientE2ETests.cs (hunk 3 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/service/DigitalTwinClientE2ETests.cs:69]: No, it doesn't. I can create a helper for this as well.

`Message=ErrorCode:InvalidDigitalTwinPatch;Error in path: targetTemperature. 'path' string must be non-null, begin with '/' and have atleast the component name.`

File: iothub/service/src/DigitalTwin/Extensions.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+using System;
+using System.Globalization;
+
+namespace Microsoft.Azure.Devices
+{
+    internal static class Extensions
+    {
+        public const string ParameterNullErrorMessageFormat = "The parameter named {0} can't be null.";
+        public const string ParameterNullWhiteSpaceErrorMessageFormat = "The parameter named {0} can't be null, empty string or white space.";
```
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/Extensions.cs:12]: just empty not empty string

File: iothub/service/src/DigitalTwin/Generated/DigitalTwin.cs [hunk 7]
```diff
@@ -0,0 +1,803 @@
...
+        /// <return>
+        /// A response object containing the response body and response headers.
+        /// </return>
+        public async Task<HttpOperationResponse<string,DigitalTwinGetHeaders>> GetDigitalTwinWithHttpMessagesAsync(string id, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (id == null)
+            {
+                throw new ValidationException(ValidationRules.CannotBeNull, "id");
+            }
+            string apiVersion = "2020-09-30";
```
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/Generated/DigitalTwin.cs:81]: Do we need to parameterize this? We should have it specified in one location, so if/when we change it it isn't in many places.

File: e2e/test/Helpers/digitaltwins/ThermostatTwin.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (drwill-ms) [inline e2e/test/Helpers/digitaltwins/ThermostatTwin.cs:1]: Probably should be under a folder called `models`, not helpers.

File: iothub/device/samples/DigitalTwinDeviceSamples/PnpHelper/PnpHelper.cs [hunk 9]
```diff
@@ -153,6 +154,19 @@ public static (bool, T) GetPropertyFromTwin<T>(TwinCollection collection, string
             return (false, default);
         }
 
+        /// <summary>
+        /// Create a property patch for component-level property updates.
+        /// </summary>
+        /// <param name="propertyKeyValuePairs">The dictionary of property key values pairs that are to be updated.</param>
+        /// <returns>The dictionary containing the property key value pairs for a component-level property update.</returns>
+        public static Dictionary<string, object> CreatePatchForComponentUpdate(Dictionary<string, object> propertyKeyValuePairs)
```
Comment (drwill-ms) [inline iothub/device/samples/DigitalTwinDeviceSamples/PnpHelper/PnpHelper.cs:162]: I don't understand why this method exists.

File: iothub/service/src/DigitalTwin/Authentication/IotServiceClientCredentials.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.Devices.Authentication
+{
+    internal abstract class IotServiceClientCredentials : ServiceClientCredentials
+    {
+        /// <summary>
+        /// Add a sas token to the outgoing http request, then send it to the next pipeline segment
```
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/Authentication/IotServiceClientCredentials.cs:16]: SAS should be capitalized, I think

File: iothub/service/src/DigitalTwin/DigitalTwinClient.cs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            _protocolLayer = new DigitalTwin(_client);
+        }
+
+        /// <summary>
+        /// Gets a strongly-typed digital twin asynchronously.
+        /// </summary>
+        /// <param name="digitalTwinId">The Id of the digital twin.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The application/json digital twin and the http response.</returns>
+        public async Task<HttpOperationResponse<T, DigitalTwinGetHeaders>> GetAsync<T>(string digitalTwinId, CancellationToken cancellationToken = default)
```
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/DigitalTwinClient.cs:46]: If PnP adds other methods to this client, like for getting models and such, `GetAsync` is going to not be obvious anymore. Any reason not to just use `GetDigitalTwinAsync`?

**Review by drwill-ms — APPROVED** [review]

File: e2e/test/iothub/service/DigitalTwinClientE2ETests.cs (hunk 3 above)
Comment (drwill-ms) [reply e2e/test/iothub/service/DigitalTwinClientE2ETests.cs:69]: We should document it in the very least. A helper might be an overkill. I'd almost rather the client code look for a slash at the beginning, and prepend one. I'm curious what the Azure SDK team would suggest.

File: e2e/test/Helpers/digitaltwins/ThermostatTwin.cs (hunk 8 above)
Comment (abhipsaMisra) [reply e2e/test/Helpers/digitaltwins/ThermostatTwin.cs:1]: sure!

File: e2e/test/iothub/service/DigitalTwinClientE2ETests.cs (hunk 3 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/service/DigitalTwinClientE2ETests.cs:69]: the client code is will get the serialized patch, so for it to inspect it and append a "/" if missing, we would have to deserialize and serialize it again. 
I can add it to the update utility, to always append a "/", but that might complicate it when we have property within a component (we might need another overload).

A helper might be the simplest option here.

File: iothub/device/samples/DigitalTwinDeviceSamples/PnpHelper/PnpHelper.cs (hunk 9 above)
Comment (abhipsaMisra) [reply iothub/device/samples/DigitalTwinDeviceSamples/PnpHelper/PnpHelper.cs:162]: Discussed offline. Also, this is pending recommendation from sdk board.

File: e2e/test/iothub/service/DigitalTwinClientE2ETests.cs (hunk 3 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/service/DigitalTwinClientE2ETests.cs:69]: Discussed offline. Also, this is pending recommendation from sdk board.

**Pull request merged**